crates/matrix-sdk-ui/src/room_list/room_list.rs

```diff
@@ -14,30 +14,78 @@
 
 use std::future::ready;
 
+use eyeball::{shared::Observable, Subscriber};
 use eyeball_im::{Vector, VectorDiff};
-use futures_util::Stream;
+use futures_util::{pin_mut, Stream, StreamExt};
 use matrix_sdk::{RoomListEntry, SlidingSync, SlidingSyncList};
+use tokio::{spawn, task::JoinHandle};
 
-use super::Error;
+use super::{Error, State};
 
+/// A `RoomList` represents a list of rooms, from a [`RoomListService`].
 #[derive(Debug)]
 pub struct RoomList {
     sliding_sync_list: SlidingSyncList,
+    loading_state: Observable<RoomListLoadingState>,
+    loading_state_task: JoinHandle<()>,
+}
+
+impl Drop for RoomList {
+    fn drop(&mut self) {
+        self.loading_state_task.abort();
+    }
 }
 
 impl RoomList {
     pub(super) async fn new(
         sliding_sync: &SlidingSync,
         sliding_sync_list_name: &str,
+        room_list_service_state: Subscriber<State>,
     ) -> Result<Self, Error> {
+        let sliding_sync_list = sliding_sync
+            .on_list(sliding_sync_list_name, |list| ready(list.clone()))
+            .await
+            .ok_or_else(|| Error::UnknownList(sliding_sync_list_name.to_owned()))?;
+        let loading_state = Observable::new(RoomListLoadingState::NotLoaded);
+
         Ok(Self {
-            sliding_sync_list: sliding_sync
-                .on_list(sliding_sync_list_name, |list| ready(list.clone()))
-                .await
-                .ok_or_else(|| Error::UnknownList(sliding_sync_list_name.to_owned()))?,
+            sliding_sync_list: sliding_sync_list.clone(),
+            loading_state: loading_state.clone(),
+            loading_state_task: spawn(async move {
+                pin_mut!(room_list_service_state);
+
+                // As soon as `RoomListService` changes its state, if it isn't
+                // `Terminated` nor `Error`, we know we have fetched something,
+                // so the room list is loaded.
+                while let Some(state) = room_list_service_state.next().await {
+                    use State::*;
+
+                    match state {
+                        Terminated { .. } | Error { .. } | Init => (),
+                        SettingUp | Running => break,
+                    }
+                }
+
+                // Let's jump from `NotLoaded` to `Loaded`.
+                let maximum_number_of_rooms = sliding_sync_list.maximum_number_of_rooms();
+
+                loading_state.set(RoomListLoadingState::Loaded { maximum_number_of_rooms });
+
+                // Wait for updates on the maximum number of rooms to update again.
+                while let Some(maximum_number_of_rooms) =
+                    sliding_sync_list.maximum_number_of_rooms_stream().next().await
```

Comment:
Does creating the stream cause some overhead? I'd hoist the `maximum_number_of_rooms_stream()` call, just in case ^^

Reply:
Ah, good catch!